Source/WebCore/Modules/permissions/PermissionStatus.h

```diff
@@ -28,40 +28,47 @@
 #include "ActiveDOMObject.h"
 #include "ClientOrigin.h"
 #include "EventTarget.h"
+#include "Page.h"
 #include "PermissionDescriptor.h"
 #include "PermissionName.h"
 #include "PermissionObserver.h"
+#include "PermissionQuerySource.h"
 #include "PermissionState.h"
+#include "ScriptExecutionContextIdentifier.h"
+#include <atomic>
+#include <wtf/ThreadSafeRefCounted.h>
 
 namespace WebCore {
 
 class ScriptExecutionContext;
 
-class PermissionStatus final : public PermissionObserver, public ActiveDOMObject, public RefCounted<PermissionStatus>, public EventTarget  {
+class PermissionStatus final : public PermissionObserver, public ActiveDOMObject, public ThreadSafeRefCounted<PermissionStatus>, public EventTarget  {
     WTF_MAKE_ISO_ALLOCATED(PermissionStatus);
 public:
-    static Ref<PermissionStatus> create(ScriptExecutionContext&, PermissionState, const PermissionDescriptor&);
+    static Ref<PermissionStatus> create(ScriptExecutionContext&, PermissionState, const PermissionDescriptor&, PermissionQuerySource, Page*);
     ~PermissionStatus();
 
     PermissionState state() const { return m_state; }
     PermissionName name() const { return m_descriptor.name; }
 
-    using RefCounted::ref;
-    using RefCounted::deref;
+    using ThreadSafeRefCounted<PermissionStatus>::ref;
+    using ThreadSafeRefCounted<PermissionStatus>::deref;
 
     using PermissionObserver::weakPtrFactory;
     using PermissionObserver::WeakValueType;
     using PermissionObserver::WeakPtrImplType;
 
 private:
-    PermissionStatus(ScriptExecutionContext&, PermissionState, const PermissionDescriptor&);
+    PermissionStatus(ScriptExecutionContext&, PermissionState, const PermissionDescriptor&, PermissionQuerySource, Page*);
+    void createAssociatedMainThreadPermissionObserver(ScriptExecutionContext&);
 
     // PermissionObserver
     PermissionState currentState() const final { return m_state; }
     void stateChanged(PermissionState) final;
     const ClientOrigin& origin() const final { return m_origin; }
     const PermissionDescriptor& descriptor() const final { return m_descriptor; }
-    const ScriptExecutionContext* context() const final { return ActiveDOMObject::scriptExecutionContext(); }
+    PermissionQuerySource source() const final { return m_source; }
+    Page* page() const final { return m_page.get(); }
```

Comment:
Something is weird with the design here. Seems awkward for a PermissionStatus to have a pointer to the Page, given that the Page is a main thread object and PermissionStatus can live on worker threads.

This m_page data member and page() getter are foot-guns. Or is m_page null when PermissionStatus is not a main thread object?

Reply:
Oh that's a good point. The reason for this design is that in WebPermissionController::permissionChanged, we use `observer->page()`. And by ensuring that both PermissionStatus and MainThreadPermissionObserver have a page() method, we can keep the code in WebPermissionController::permissionChanged the same. The m_page data member held by PermissionStatus comes from Permissions::query() where the page is passed in as an argument to the PermissionStatus constructor. But that page is not null, it's the page that a worker context has (the one with an empty Chrome client and such). I can change this so that m_page is null when PermissionStatus is not a main thread object and then this would be fine right?